Karma config improvements

symbolist opened the pull request:
https://openedx.atlassian.net/browse/TNL-4398

This PR contains the following changes:
1. Moves all the logic from the individual Karma config files to `karma.common.conf.js`. The config files just declare a list of files now.
2. When using RequireJS for running a suite, individual files do not need to be listed in the Karma config. If we just specify the directories, Karma will serve them correctly.
3. Moves the jasmine addons to `common/js/spec_helpers` and if possible loads them via RequireJS before loading the spec files. Then the spec files can just assume those helpers exist. For suites which do not use RequireJS for running the specs, these addons are included via script tags in the html.
4. Deletes the js-test-tool configuration files.
5. Excludes the translation files from coverage tracking.

symbolist commented:
jenkins run js

symbolist commented:
@ehteshamkafeel @andy-armstrong There is one thing I am wondering about. Given point 2, we can: 
1. reduce `libraryFiles` to ['**/*.js''] and thus remove `fixtureFiles`.
2. Merge `sourceFiles` and `specFiles` into one `filesForCoverage`.

The current config is more explicit and will encourage developers to put the files in the right places. This approach described above will be simpler to manage. Preferences?

Note that I am still working on adding documentation to these files.

symbolist commented:
@ehteshamkafeel Can you also review this?

symbolist commented:
One more thing: the following files are also no longer included in coverage:
```
[   'cms/static/coffee/spec/main.js',
    'cms/static/coffee/spec/main_squire.js',
    'common/static/common/js/common_libraries.js',
    'common/static/common/js/karma.common.conf.js',
    'common/static/common/js/spec_helpers/ajax_helpers.js',
    'common/static/common/js/spec_helpers/page_helpers.js',
    'common/static/common/js/spec_helpers/spec_helpers.js',
    'common/static/common/js/spec_helpers/template_helpers.js',
    'common/static/common/js/spec_helpers/view_helpers.js',
    'common/static/common/js/vendor/backbone.js',
    'common/static/common/js/vendor/picturefill.js',
    'common/static/common/js/vendor/underscore.js',
    'common/static/common/js/vendor/underscore.string.js',
    'lms/static/edx-pattern-library/js/afontgarde.js',
    'lms/static/edx-pattern-library/js/edx-icons.js',
    'lms/static/edx-pattern-library/js/modernizr-custom.js',
    'lms/static/edx-pattern-library/js/select-replace.js',
    'lms/static/edx-ui-toolkit/js/disclosure/disclosure-view.js',
    'lms/static/edx-ui-toolkit/js/dropdown-menu/dropdown-menu-view.js',
    'lms/static/edx-ui-toolkit/js/pagination/paging-collection.js',
    'lms/static/edx-ui-toolkit/js/utils/constants.js',
    'lms/static/edx-ui-toolkit/js/utils/global-loader.js',
    'lms/static/edx-ui-toolkit/js/utils/html-utils.js',
    'lms/static/edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers.js',
    'lms/static/edx-ui-toolkit/js/utils/spec-helpers/spec-helpers.js',
    'lms/static/edx-ui-toolkit/js/utils/string-utils.js']
```

andy-armstrong commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -180,10 +178,13 @@ requirejs.config({
         },
         "jasmine-imagediff": {},
-        "jasmine-stealth": {
+        "common/js/spec_helpers/jasmine-extensions": {
```
I don't like having vendor libraries under ``spec_helpers`` as that's a directory of our own code. Can it just live in the regular vendor directory? Ideally it would be installed via npm, but this one doesn't seem to be available.

symbolist replied:
@andy-armstrong 

1. `jasmine-extensions.js` and `jasmine-waituntil.js` were written by us.
2. The upstream `jasmine-stealth` is not compatible with Jasmine 2.0. This file contains only one helper from it that is needed in edx-platform.

andy-armstrong replied:
Ah, my mistake. Thanks for explaining!

andy-armstrong commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -180,10 +178,13 @@ requirejs.config({
+        "common/js/spec_helpers/jasmine-extensions": {
+            deps: ["jquery"]
+        },
+        "common/js/spec_helpers/jasmine-stealth": {
```
This one is available in npm: https://www.npmjs.com/package/jasmine-stealth

andy-armstrong commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -180,10 +178,13 @@ requirejs.config({
             deps: ["underscore", "underscore.string"]
         },
-        "jasmine-waituntil": {
+        "common/js/spec_helpers/jasmine-waituntil": {
```
This one isn't in npm, but there is a similar sounding library: https://www.npmjs.com/package/jasmine-waits-for-callback

andy-armstrong replied:
Never mind. That library seems very old and probably predates Jasmine 2.

ehteshamkafeel commented on common/static/common/js/karma.common.conf.js:
```diff
@@ -102,27 +124,60 @@ function junitSettings(config) {
-            pattern = path.join(appRoot, '/common/static/' +
-                pattern.replace(/^xmodule_js\/common_static\//, ''));
+        var pattern = _.isObject(file) ? file.pattern : file;
+        if (normalize_func) {
```
Please make it camelCase

ehteshamkafeel commented on common/static/common/js/karma.common.conf.js:
```diff
@@ -102,27 +124,60 @@ function junitSettings(config) {
+    });
+
+    return filesForCoverage;
+};
```
This logic can be better written if we make a normalize function here in this file a default one.

```javascript
var defaultNormalizeFunc = function (***)  {
  if (pattern.match(/^common\/js/))  {
     pattern = path.join(appRoot, '/common/static/' + pattern);
  } else if (pattern.match(/^xmodule_js\/common_static/)) {
     pattern = path.join(appRoot, '/common/static/' +
     pattern.replace(/^xmodule_js\/common_static\//, ''));
   }
}
```

and then in the `normalizePathsForCoverage` body

```javascript

var normalizePathsForCoverage = function(files, normalizeFn) {
  normalizeFn = normalizeFn || defaultNormalizeFunc;

  // Later
  pattern = normalizeFn(***);
}

ehteshamkafeel commented on common/static/common/js/karma.common.conf.js:
```diff
@@ -102,27 +124,60 @@ function junitSettings(config) {
+ * @param {Bool} enable
+ * @return {Object}
+ */
+var setNocache = function (files, enable) {
```
a better name would be `shouldCache` for param.

ehteshamkafeel commented on lms/static/js/spec/main.js:
```diff
@@ -776,11 +773,18 @@
+        'common/js/spec_helpers/jasmine-waituntil'
+    ];
 
+    require(specHelpers, function() {
```
can't we just prepend spec helper files to testFiles? They are getting load serially after all.

ehteshamkafeel replied:
Since these are libraries we can move it to edx-ui-toolkit some time in future.

andy-armstrong commented on cms/static/karma_cms.conf.js:
```diff
@@ -1,142 +1,48 @@
-var _ = require('underscore');
 var configModule = require(path.join(__dirname, '../../common/static/common/js/karma.common.conf.js'));
 
-var libraryFiles = [
```
Yes! 💯  for this clean up!

andy-armstrong commented on common/static/common/js/karma.common.conf.js:
```diff
@@ -102,27 +124,60 @@ function junitSettings(config) {
+var normalizePathsForCoverage = function(files, normalize_func) {
+    var filesForCoverage = {};
 
     files.forEach(function (file) {
```
I'm finding this logic hard to read. Can you name this anonymous function and give it a docstring.

andy-armstrong commented on common/static/common/js/karma.common.conf.js:
```diff
@@ -235,8 +290,52 @@ var getConfig = function (config, useRequireJs) {
+    // With nocache=true, Karma always serves the latest files from disk.
+    // However, that prevents coverage tracking from working.
+    // So we only set it if coverage tracking is off.
+    setNocache(files, !data.config.coverage);
```
Why do we want nocache on at all? It seems that we would only want this in debug mode.

symbolist replied:
@andy-armstrong If `nocache=false`, Karma reads files once and keeps serving them from its cache. So if I make a change to a file Karma will need to be restarted for the changes to be reflected. In dev mode, we want changes to be reflected on a page refresh. And in run mode the browser will request a file only once anyway. So as long as we do not need coverage we set it to `true`.

andy-armstrong commented on lms/static/js/spec/main_requirejs_coffee.js:
```diff
@@ -4,9 +4,9 @@
-            "modernizr": "xmodule_js/common_static/edx-pattern-library/js/modernizr-custom",
-            "afontgarde": "xmodule_js/common_static/edx-pattern-library/js/afontgarde",
-            "edxicons": "xmodule_js/common_static/edx-pattern-library/js/edx-icons",
+            "modernizr": "edx-pattern-library/js/modernizr-custom",
```
I'm so happy to not have to use the bizarre common_static path.

andy-armstrong commented on lms/static/karma_lms.conf.js:
```diff
@@ -1,155 +1,62 @@
-var preprocessors = configModule.getPreprocessorObject(_.flatten([sourceFiles, specFiles]));
+var files = {
+    libraryFilesToInclude: [
+        {pattern: 'xmodule_js/common_static/js/vendor/jquery.event.drag-2.2.js', included: true},
```
Can we load these from {{common}} rather than {{xmodule_js/common_static}} now?

ehteshamkafeel replied:
yeah they can be made to load from common.

ehteshamkafeel commented:
@symbolist This looks great. Much cleaner no doubt. Just left few comments.

Initially we decided to just copy files as is from `yml` files for karma config. The purpose to list individual files was that some files weren't needed for tests and they were included by karma. Of course there is no harm but using this pattern `js/**/*.js` crashed karma runner once.

andy-armstrong commented:
@symbolist Fantastic clean up, thanks! I love how much duplicated code you removed, and how much simpler it will be to maintain going forwards. I have just a few minor comments.

One thing: I would like to see the ``common/js/spec_helpers`` included in the coverage. It seems similar to how we keep the spec files themselves, so we can understand what tests are running.

symbolist commented:
@andy-armstrong @ehteshamkafeel I have addressed the feedback.

1. `normalizePathsForCoverage` has been refactored.
2. `xmodule_js` prefix from paths of `edx-ui-toolkit`, `edx-pattern-library` and `common` have been removed.
3. Patterns for files in `edx-ui-toolkit`, `edx-pattern-library` and `common` have been moved to `karma.common.conf.js` and are shared by all the cms/lms suites.
4. Some documentation has been added.
5. Coverage of files in `common/js/spec_helpers` is being tracked. It is under the `spec_helpers` package. So overall the translation files and the following files have been removed from coverage tracking:
```
[   'cms/static/coffee/spec/main.js',
    'cms/static/coffee/spec/main_squire.js',
    'common/static/common/js/common_libraries.js',
    'common/static/common/js/jasmine.common.conf.js',
    'common/static/common/js/karma.common.conf.js',
    'common/static/common/js/vendor/backbone.js',
    'common/static/common/js/vendor/picturefill.js',
    'common/static/common/js/vendor/underscore.js',
    'common/static/common/js/vendor/underscore.string.js',
    'lms/static/edx-pattern-library/js/afontgarde.js',
    'lms/static/edx-pattern-library/js/edx-icons.js',
    'lms/static/edx-pattern-library/js/modernizr-custom.js',
    'lms/static/edx-pattern-library/js/select-replace.js',
    'lms/static/edx-ui-toolkit/js/disclosure/disclosure-view.js',
    'lms/static/edx-ui-toolkit/js/dropdown-menu/dropdown-menu-view.js',
    'lms/static/edx-ui-toolkit/js/pagination/paging-collection.js',
    'lms/static/edx-ui-toolkit/js/utils/constants.js',
    'lms/static/edx-ui-toolkit/js/utils/global-loader.js',
    'lms/static/edx-ui-toolkit/js/utils/html-utils.js',
    'lms/static/edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers.js',
    'lms/static/edx-ui-toolkit/js/utils/spec-helpers/spec-helpers.js',
    'lms/static/edx-ui-toolkit/js/utils/string-utils.js']
```

Can you review the last 4 commits?

andy-armstrong commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -27,9 +27,9 @@ requirejs.config({
+        "underscore": "common/js/vendor/underscore",
+        "underscore.string": "common/js/vendor/underscore.string",
+        "backbone": "common/js/vendor/backbone",
         "backbone.associations": "xmodule_js/common_static/js/vendor/backbone-associations-min",
```
Do these remaining references to ``xmodule_js/common_static`` still need to be used? Can they be switched to ``common`` as well? Don't worry about it for this PR if you need to just wrap this up.

symbolist replied:
@andy-armstrong The paths are relative to the relevant `static` directory of suite. We do have symlinks to `edx-ui-toolkit`, `edx-pattern-library` and `common` in these directories and so the paths could be changed easily.

I noticed today that Karma has a `proxy` option which can be used to rewrite URLs. We can use that to get rid of these references too. But I think that will let us use the full paths instead of the symlinks too (and thus get rid of the path rewriting that is being done for coverage). So it would make sense to do all that work in one go.

andy-armstrong replied:
Right. I used proxies in the UI Toolkit's Karma configuration for exactly this:

https://github.com/edx/edx-ui-toolkit/blob/master/test/karma.conf.js#L28

andy-armstrong replied:
I've created a JIRA story to track this idea:

https://openedx.atlassian.net/browse/FEDX-181

andy-armstrong commented:
👍 Fantastic stuff, @symbolist. We're going to miss you!

symbolist commented:
jenkins run js
